src/txdb.cpp

```diff
@@ -68,10 +69,30 @@ uint256 CCoinsViewDB::GetBestBlock() const {
     return hashBestChain;
 }
 
+std::vector<uint256> CCoinsViewDB::GetHeadBlocks() const {
+    std::vector<uint256> vhashHeadBlocks;
+    if (!db.Read(DB_HEAD_BLOCKS, vhashHeadBlocks))
+        return std::vector<uint256>();
+    return vhashHeadBlocks;
+}
+
 bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
     CDBBatch batch(db);
     size_t count = 0;
     size_t changed = 0;
+    size_t batch_size = (size_t)GetArg("-dbbatchsize", nDefaultDbBatchSize);
+    if (!hashBlock.IsNull()) {
+        uint256 old_tip = GetBestBlock();
+        if (old_tip.IsNull()) {
+            // We may be in the middle of replaying.
+            std::vector<uint256> old_heads = GetHeadBlocks();
+            if (old_heads.size() >= 2) old_tip = old_heads[1];
+        }
+        // In the first batch, mark the database as being in the middle of a
+        // transition from old_tip to hashBlock.
+        batch.Erase(DB_BEST_BLOCK);
+        batch.Write(DB_HEAD_BLOCKS, std::vector<uint256>{hashBlock, old_tip});
```

Comment:
It's for future compatibility with a design where entries are continuously written to disk in the background, which may lead to multiple partially-applied reorgs to co-exist on disk.